core/stat/base/leap_array.go

```diff
@@ -49,28 +49,25 @@ type AtomicBucketWrapArray struct {
 	data   []*BucketWrap
 }
 
-// New AtomicBucketWrapArray with initializing field data
-// Default, automatically initialize each BucketWrap
-// len: length of array
-// bucketLengthInMs: bucket length of BucketWrap
-// generator: generator to generate bucket
-func NewAtomicBucketWrapArray(len int, bucketLengthInMs uint32, generator BucketGenerator) *AtomicBucketWrapArray {
+func NewAtomicBucketWrapArrayWithTime(len int, bucketLengthInMs uint32, now uint64, generator BucketGenerator) *AtomicBucketWrapArray {
 	ret := &AtomicBucketWrapArray{
 		length: len,
 		data:   make([]*BucketWrap, len),
 	}
 
 	// automatically initialize each BucketWrap
 	// tail BucketWrap of data is initialized with current time
-	startTime := calculateStartTime(util.CurrentTimeMillis(), bucketLengthInMs)
-	for i := len - 1; i >= 0; i-- {
+	interval := uint64(bucketLengthInMs * uint32(len))
+	arrayStartTime := now - (now % interval)
```

Comment:
The startTime should be `now - now % bucketLengthInMs`?

For example (in related test cases), sampleCount=10, bucketLength=200ms, now=1596199317921, then the first (and oldest) bucket of created data structure should be 1596199317800